engine/common/requester/engine.go

```diff
@@ -426,6 +439,15 @@ func (e *Engine) onEntityResponse(originID flow.Identifier, res *messages.Entity
 		delete(needed, entityID)
 		delete(e.items, entityID)
 
+		if item.checkIntegrity {
+			actualEntityID := entity.ID()
+			// validate that we got correct entity, exactly what we were expecting
+			if entityID != actualEntityID {
+				return engine.NewInvalidInputErrorf("could not validate response body for requested entity, expected: %v, actual: %v",
+					entityID, actualEntityID)
+			}
+		}
+
```

Comment:
Why? We got the response, we remove it from pending items either way, even if we got invalid entry.

Reply:
🤔 Not sure I follow your logic. Consider the case where a consensus node is missing a block with ID `b`. So the node sends out a request to evil Eve: 
```
EntityRequest{
   Nonce = 21,
   EntityIDs = [b]
}
``` 
Eve responds with:
```
EntityRequest{
   Nonce = 21,
   EntityIDs = [b]
   Blobs = [some other block]
}
``` 

I think your logic will remove `entityID` from the list of entities to be requested: https://github.com/onflow/flow-go/blob/d0459d294c7dbae5bfb766603507d00b69c9ca18/engine/common/requester/engine.go#L440 
* `entityID` is the ID of the entity that `Eve` _claims_ to send us. Not the ID of the entity she actually sent

I think, Eve can make our node stop requesting needed entities without actually fulfilling the request. This would not be BFT.

Reply:
specifically, I am suggesting

1. check that the provided entity actually matches the stated ID
2. if that is the case (or we are trusting the origin)

```golang
		if item.checkIntegrity {
			actualEntityID := entity.ID()
			// validate that we got correct entity, exactly what we were expecting
			if entityID != actualEntityID {
				e.log.Error().
					Hex("origin", logging.ID(originID)).
					Hex("stated_entity_id", logging.ID(entityID)).
					Hex("provided_entity", logging.ID(actualEntityID)).
					Msg("provided entity does not match stated ID")
				continue
			}
		}

		// remove from needed items and pending items
		delete(needed, entityID)
		delete(e.items, entityID)
```

Reply:
right, I somewhy thought that it's client based so we are deleting entity only for one client